File: experiments/evm_feasibility/p256_reference.go
// Pinned direct Go oracle for Taraxa's Cacti P-256 precompile.
package main

import (
	"crypto/elliptic"
	"encoding/hex"
	"encoding/json"
	"math/big"
	"os"

	"github.com/Taraxa-project/taraxa-evm/common"
	"github.com/Taraxa-project/taraxa-evm/core/vm"
)

type p256Row struct {
	Name        string `json:"name"`
	Input       string `json:"input"`
	RequiredGas uint64 `json:"required_gas"`
	Output      string `json:"output"`
	Error       string `json:"error"`
}

func p256OracleRow(precompile vm.PrecompiledContract, name string, input []byte) p256Row {
	frame := vm.CallFrame{Input: input}
	output, err := precompile.Run(frame, nil)
	errorText := ""
	if err != nil {
		errorText = err.Error()
	}
	return p256Row{
		Name: name, Input: hex.EncodeToString(input),
		RequiredGas: precompile.RequiredGas(frame, nil), Output: hex.EncodeToString(output), Error: errorText,
	}
}

func clone(input []byte) []byte { return append([]byte(nil), input...) }

func replace(input []byte, start int, value []byte) []byte {
	result := clone(input)
	copy(result[start:start+len(value)], value)
	return result
}

func leftPad32(value []byte) []byte {
	result := make([]byte, 32)
	copy(result[32-len(value):], value)
	return result
}

func main() {
	address := common.BytesToAddress([]byte{0x01, 0x00})
	precompile := vm.PrecompiledContractsCacti.Get(&address)
	if precompile == nil {
		panic("Cacti P-256 precompile is absent")
	}
	valid, err := hex.DecodeString("4cee90eb86eaa050036147a12d49004b6b9c72bd725d39d4785011fe190f0b4da73bd4903f0ce3b639bbbf6e8e80d16931ff4bcf5993d58468e8fb19086e8cac36dbcd03009df8c59286b162af3bd7fcc0450c9aa81be5d10d312af6c66b1d604aebd3099c618202fcfe16ae7770b0c49ab5eadf74b754204a3bb6060e44eff37618b065f9832de4ca6ca971a7a1adc826d0f7c00181a5fb2ddf79ae00b4e10e")
	if err != nil || len(valid) != 160 {
		panic("valid P-256 vector")
	}

	order := elliptic.P256().Params().N
	s := clone(valid[64:96])
	highS := leftPad32(new(big.Int).Sub(new(big.Int).Set(order), new(big.Int).SetBytes(s)).Bytes())
	zero := make([]byte, 32)
	curveOrder := leftPad32(order.Bytes())
	fieldPrime := leftPad32(elliptic.P256().Params().P.Bytes())

	inputs := []struct {
		name  string
		input []byte
	}{
		{"empty", nil},
		{"one-byte", []byte{1}},
		{"zero-159", make([]byte, 159)},
		{"zero-160", make([]byte, 160)},
		{"zero-161", make([]byte, 161)},
		{"valid", valid},
		{"valid-truncated", clone(valid[:159])},
		{"valid-trailing", append(clone(valid), 0)},
		{"wrong-message", append([]byte{valid[0] ^ 1}, valid[1:]...)},
		{"high-s-valid", replace(valid, 64, highS)},
		{"r-zero", replace(valid, 32, zero)},
		{"r-order", replace(valid, 32, curveOrder)},
		{"s-zero", replace(valid, 64, zero)},
		{"s-order", replace(valid, 64, curveOrder)},
		{"public-x-zero", replace(valid, 96, zero)},
		{"public-y-zero", replace(valid, 128, zero)},
		{"public-x-field-prime", replace(valid, 96, fieldPrime)},
		{"public-y-field-prime", replace(valid, 128, fieldPrime)},
	}
	rows := make([]p256Row, 0, len(inputs))
	for _, input := range inputs {
		rows = append(rows, p256OracleRow(precompile, input.name, input.input))
	}
	if err := json.NewEncoder(os.Stdout).Encode(map[string]any{"p256": rows}); err != nil {
		panic(err)
	}
}

File: experiments/evm_feasibility/p256_reference.py
#!/usr/bin/env python3
"""Reproduce the bounded dual-pin Go Cacti P-256 corpus.

The default mode verifies checked-in bytes and hashes. ``--record`` replaces
the three owned fixture files only after both immutable revisions agree.
"""

import argparse
import hashlib
import json
import os
import pathlib
import subprocess
import tempfile

from reference import REVISIONS, ROOT

HERE = pathlib.Path(__file__).resolve().parent
FIXTURES = HERE / "fixtures" / "p256"


def fail(message):
    raise RuntimeError(message)


def checked_output(command, **kwargs):
    try:
        return subprocess.check_output(command, **kwargs)
    except subprocess.CalledProcessError as error:
        fail(f"command failed ({error.returncode}): {' '.join(map(str, command))}")


def validate_artifact(label, data):
    try:
        document = json.loads(data)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        fail(f"{label} emitted invalid JSON: {error}")
    rows = document.get("p256")
    if not isinstance(rows, list) or not rows:
        fail(f"{label} omitted P-256 rows")
    names = set()
    for index, row in enumerate(rows):
        if not isinstance(row, dict):
            fail(f"{label} row {index} is not an object")
        if set(row) != {"name", "input", "required_gas", "output", "error"}:
            fail(f"{label} row {index} has unexpected fields")
        name = row["name"]
        if not isinstance(name, str) or name in names:
            fail(f"{label} duplicate or invalid row name: {name!r}")
        names.add(name)
        for field in ("input", "output"):
            try:
                bytes.fromhex(row[field])
            except (TypeError, ValueError) as error:
                fail(f"{label} row {index} invalid {field}: {error}")
        if row["required_gas"] != 6900:
            fail(f"{label} row {index} has wrong gas")
        if row["error"] != "":
            fail(f"{label} row {index} unexpectedly errors")
        if row["output"] not in ("", "00" * 31 + "01"):
            fail(f"{label} row {index} has non-boolean output")
    needed = {
        "empty",
        "zero-159",
        "zero-160",
        "zero-161",
        "valid",
        "valid-truncated",
        "valid-trailing",
        "wrong-message",
        "high-s-valid",
        "r-zero",
        "r-order",
        "s-zero",
        "s-order",
        "public-x-zero",
        "public-y-zero",
        "public-x-field-prime",
        "public-y-field-prime",
    }
    missing = needed - names
    if missing:
        fail(f"{label} omitted required cases: {sorted(missing)}")
    by_name = {row["name"]: row for row in rows}
    if by_name["valid"]["output"] != "00" * 31 + "01":
        fail(f"{label} valid signature did not return true")
    if by_name["high-s-valid"]["output"] != "00" * 31 + "01":
        fail(f"{label} high-S signature was not accepted")
    for name, row in by_name.items():
        if name not in ("valid", "high-s-valid") and row["output"] != "":
            fail(f"{label} invalid case returned true: {name}")


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--record", action="store_true")
    args = parser.parse_args()

    source = pathlib.Path(os.environ.get("TARAXA_EVM_SOURCE", ROOT / "submodules/taraxa-evm"))
    exporter = (HERE / "p256_reference.go").read_bytes()
    artifacts = {}
    for label, revision in REVISIONS.items():
        with tempfile.TemporaryDirectory(prefix="rustaxa-p256-") as directory:
            tree = pathlib.Path(directory)
            archive = checked_output(["git", "-C", str(source), "archive", revision])
            subprocess.run(["tar", "-x", "-C", directory], input=archive, check=True)
            command = tree / "cmd/p256_reference"
            command.mkdir(parents=True)
            (command / "p256_reference.go").write_bytes(exporter)
            data = checked_output(
                ["go", "run", "-mod=readonly", "./cmd/p256_reference"], cwd=tree
            )
            validate_artifact(label, data)
            artifacts[label] = data

    if artifacts["public"] != artifacts["local"]:
        fail("pinned Go P-256 outputs disagree")

    manifest = {
        "schema": 1,
        "references": REVISIONS,
        "go_version": checked_output(["go", "version"], text=True).strip(),
        "scope": (
            "Direct Cacti address 0x0100 RequiredGas/Run over exact-length, "
            "signature-scalar and public-key boundaries; no EVM admission, registry, "
            "fork activation, frame, state or production-routing claim"
        ),
        "sha256": {
            label: hashlib.sha256(data).hexdigest() for label, data in artifacts.items()
        },
        "exporter_sha256": hashlib.sha256(exporter).hexdigest(),
    }
    manifest_bytes = (json.dumps(manifest, indent=2) + "\n").encode()
    expected = {
        "public.json": artifacts["public"],
        "local.json": artifacts["local"],
        "manifest.json": manifest_bytes,
    }
    if args.record:
        FIXTURES.mkdir(parents=True, exist_ok=True)
        for name, data in expected.items():
            (FIXTURES / name).write_bytes(data)
        return
    for name, data in expected.items():
        path = FIXTURES / name
        if not path.is_file():
            fail(f"missing fixture: {path}; run with --record after review")
        if path.read_bytes() != data:
            fail(f"fixture differs: {path}; run with --record only for an intentional update")


if __name__ == "__main__":
    main()
